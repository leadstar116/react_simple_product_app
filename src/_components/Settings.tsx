import React, { useEffect, useState } from 'react'
import { Row, Col } from 'react-bootstrap'
import { useDispatch, useSelector } from 'react-redux'
import { initialState } from '../_constants/state.interface'
import { loadCurrencies, updateCurrentCurrency } from '../_helpers/settings.thunk'

const Settings = () => {
    const dispatch = useDispatch()
    const settings = useSelector((state: initialState) => state.settingsReducer)
    const [isInitialized, setIsInitialized] = useState(false)

    // Load Currencies
    useEffect(() => {
        if(isInitialized
            || settings.currencies.length)
            return
        dispatch(loadCurrencies())
        setIsInitialized(true)
    }, [dispatch, isInitialized, settings])

    const findCurrency = (code: string) => {
        return settings.currencies.find((currency) => currency.code === code)
    }
    return (
        <div className="p-2">
            <h2 className="text-center">Settings</h2>
            <div className="settings-container">
                <Row className="text-center">
                    <Col xs="12" sm="3">Currencies: </Col>
                    <Col xs="12" sm="9">
                        <select
                            name="currency"
                            id="currency"
                            className="currency-select"
                            value={settings.currentCurrency.code}
                            onChange={(e) => dispatch(updateCurrentCurrency(findCurrency(e.target.value)))}
                        >
                            {
                                settings.currencies.map((currency, index) => (
                                    <option
                                        value={currency.code}
                                        key={index}
                                    >
                                        {currency.name + "(" + currency.code + ")"}
                                    </option>
                                ))
                            }
                        </select>
                    </Col>
                </Row>
            </div>
        </div>
    )
}

export default Settings