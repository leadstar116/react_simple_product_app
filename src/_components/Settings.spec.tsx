import configureStore from 'redux-mock-store'
import React from 'react'
import { Provider } from 'react-redux'
import renderer, { ReactTestRenderer } from 'react-test-renderer'
import Settings from './Settings'
import { Currency } from '../_constants/settings.interface'

const mockStore = configureStore()

const initialState = {
    currentCurrency: {
        name: 'US Dollar',
        symbol: '$',
        symbolNative: '$',
        decimalDigits: 2,
        rounding: 0,
        code: 'USD',
        namePlural: 'US dollars',
        countries: [
            'AS',
            'BQ',
            'EC',
            'FM',
            'GU',
            'IO',
            'MH',
            'MP',
            'PR',
            'PW',
            'SV',
            'TC',
            'TL',
            'UM',
            'US',
            'VG',
            'VI',
        ],
    } as Currency,
    currencies: [] as Currency[]
}

describe('Settings Component', () => {
    let store
    let component: ReactTestRenderer

    store = mockStore({
        settingsReducer: initialState
    })

    store.dispatch = jest.fn()

    component = renderer.create(
        <Provider store={store}>
            <Settings />
        </Provider>
    )

    it('should render with given state from Redux store', () => {
        expect(component.toJSON()).toMatchSnapshot()
    })
})