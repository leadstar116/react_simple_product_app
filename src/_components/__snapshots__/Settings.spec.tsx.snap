// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings Component should render with given state from Redux store 1`] = `
<div
  className="p-2"
>
  <h2
    className="text-center"
  >
    Settings
  </h2>
  <div
    className="settings-container"
  >
    <div
      className="text-center row"
    >
      <div
        className="col-sm-3 col-12"
      >
        Currencies: 
      </div>
      <div
        className="col-sm-9 col-12"
      >
        <select
          className="currency-select"
          id="currency"
          name="currency"
          onChange={[Function]}
          value="USD"
        />
      </div>
    </div>
  </div>
</div>
`;
